### PR TITLE
#10137: Move remaining binary composite ops to ttnn

### PR DESCRIPTION
**Pull request by mcw-anasuya**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/10137)

### Problem description

Migrate composite Binary op into ttnn

### What's changed

- [x] div
- [x] div_no_nan
- [x] remainder
- [x] fmod
- [x] floor_div
- [x] logical_and_
- [x] logical_or_
- [x] logical_xor_

### Checklist
- [x] Post commit CI passes
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/eltwise/binary/binary_composite.hpp [hunk 1]
```diff
@@ -22,7 +22,7 @@ struct ExecuteBinaryCompositeOps
         const Tensor& input_tensor_b,
         const std::optional<MemoryConfig>& memory_config = std::nullopt)
         {
-            auto op_type = get_function_type0<binary_comp_op_type>();
+            auto op_type = get_function<binary_comp_op_type>();
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/binary/binary_composite.hpp:25]: something like get_binary_fn

File: ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_composite_op.hpp [hunk 2]
```diff
@@ -117,20 +150,76 @@ struct OpHandler_IsClose<BinaryCompositeOpType::ISCLOSE> {
...
 }
 
+using HandleFunctionPtr1 = Tensor (*)(const Tensor&, const Tensor&, const std::optional<MemoryConfig>&);
+using HandleFunctionPtr2 = Tensor (*)(const Tensor&, float, const std::optional<MemoryConfig>&);
+template <BinaryCompositeOpType OpType>
+auto get_function_divnonan_floordiv() -> HandleFunctionPtr1 {
+    return &OpHandler_Overload<OpType>::handle;
+}
+template <BinaryCompositeOpType OpType>
+auto get_function_divnonan_floordiv_overload() -> HandleFunctionPtr2 {
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_composite_op.hpp:210]: if these 2 fns are used by both divnonan and floordiv ops, make it common like, get_div_ops_fn, get_div_ops_fn_overload

File: ttnn/cpp/ttnn/operations/eltwise/binary/binary_pybind.hpp [hunk 3]
```diff
@@ -253,6 +253,114 @@ void bind_binary_composite_with_rtol_atol(py::module& module, const binary_opera
             py::arg("memory_config") = std::nullopt});
 }
 
+template <typename binary_operation_t>
+void bind_binary_composite_with_overload(py::module& module, const binary_operation_t& operation, const std::string& description) {
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/binary/binary_pybind.hpp:257]: can we make this bind_binary_composite_with_overload a op specific name if it used by a common set of ops? bind_div_ops ?

File: ttnn/cpp/ttnn/operations/eltwise/binary/binary_composite.hpp [hunk 4]
```diff
@@ -52,11 +52,59 @@ struct ExecuteBinaryCompositeOpsIsClose
...
         const std::optional<MemoryConfig>& memory_config = std::nullopt)
         {
-            auto op_type = get_function_type2<binary_comp_op_type>();
+            auto op_type = get_function_isclose<binary_comp_op_type>();
             return op_type(input_tensor_a, input_tensor_b, rtol, atol, equal_nan, memory_config);
         }
 };
 
+template <BinaryCompositeOpType binary_comp_op_type>
+struct ExecuteBinaryCompositeOpsOverload
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/binary/binary_composite.hpp:61]: ExecuteDivLikeOps ?

File: ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_composite_op.cpp [hunk 5]
```diff
@@ -152,4 +151,139 @@ Tensor _logical_xor(const Tensor& input_a, const Tensor& input_b, const std::opt
...
+    }
+    return result;
+}
+
+Tensor _div(const Tensor& input_a, const Tensor& input_b, bool accurate_mode, string round_mode, const std::optional<MemoryConfig>& output_mem_config) {
+    TT_FATAL((round_mode == "None" || round_mode == "trunc" || round_mode == "floor") && "Incorrect rounding mode (expected 'None', 'trunc', or 'floor')");
+    auto arch = input_a.device()->arch();
+    if (arch == tt::ARCH::WORMHOLE_B0) {
+        DataType input_dtype = input_a.get_dtype();
+        Tensor a = typecast(input_a, DataType::FLOAT32);
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_composite_op.cpp:171]: check if this points to ttnn::typecast

File: ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_composite_op.cpp [hunk 6]
```diff
@@ -152,4 +151,139 @@ Tensor _logical_xor(const Tensor& input_a, const Tensor& input_b, const std::opt
...
+        }
+        else if(round_mode == "floor"){
+            result = ttnn::floor(result);
+        }
+
+        if (accurate_mode == false) {  // If input_b is non-zero tensor
+            return typecast(result, input_dtype);
+        }
+
+        Tensor t_inf = full_like(input_a, std::numeric_limits<float>::infinity());
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_composite_op.cpp:186]: try ttnn::full_like

File: ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_composite_op.cpp [hunk 7]
```diff
@@ -152,4 +151,139 @@ Tensor _logical_xor(const Tensor& input_a, const Tensor& input_b, const std::opt
...
+        Tensor t_nan = full_like(input_a, std::nanf(""));
+        return typecast(where(
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_composite_op.cpp:188]: try using ttnn::where in all places possible

File: ttnn/ttnn/operations/binary.py [hunk 8]
```diff
@@ -65,6 +65,24 @@ def _golden_function(input_tensor_a, input_tensor_b, *args, **kwargs):
 ttnn.attach_golden_function(ttnn.eq, golden_function=_golden_function)
 
 
+def _golden_function(input_tensor_a, input_tensor_b, *args, **kwargs):
+    import torch
+
+    return torch.logical_and(input_tensor_a, input_tensor_b)
```
Comment (KalaivaniMCW) [inline ttnn/ttnn/operations/binary.py:71]: torch.logical_and_

File: ttnn/ttnn/operations/binary.py [hunk 9]
```diff
@@ -65,6 +65,24 @@ def _golden_function(input_tensor_a, input_tensor_b, *args, **kwargs):
...
+
+
+ttnn.attach_golden_function(ttnn.logical_and_, golden_function=_golden_function)
+
+
+def _golden_function(input_tensor_a, input_tensor_b, *args, **kwargs):
+    import torch
+
+    return torch.logical_or(input_tensor_a, input_tensor_b)
```
Comment (KalaivaniMCW) [inline ttnn/ttnn/operations/binary.py:80]: torch.logical_or_

File: ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_composite_op.hpp [hunk 10]
```diff
@@ -47,6 +49,9 @@ Tensor _div_no_nan(const Tensor&, const Tensor&, const std::optional<MemoryConfi
 Tensor _div_no_nan_overload(const Tensor&, float, const std::optional<MemoryConfig>&);
 Tensor _floor_div(const Tensor&, const Tensor&, const std::optional<MemoryConfig>&);
 Tensor _floor_div_overload(const Tensor&, float, const std::optional<MemoryConfig>&);
+Tensor _logical_xor_(const Tensor&, const Tensor&, const std::optional<MemoryConfig>&);
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_composite_op.hpp:52]: remove xor as it is not implemented

File: ttnn/cpp/ttnn/operations/eltwise/binary/binary_composite.hpp [hunk 11]
```diff
@@ -52,8 +52,56 @@ struct ExecuteBinaryCompositeOpsIsClose
...
+};
+
+template <BinaryCompositeOpType binary_comp_op_type>
+struct ExecuteBinaryCompositeOpsDiv
+{
+    static Tensor execute_on_worker_thread(
+        const Tensor& input_tensor_a,
+        const Tensor& input_tensor_b,
+        bool accurate_mode = false,
+        string round_mode = "None",
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/operations/eltwise/binary/binary_composite.hpp:89]: if 'string' compiles it means there is hidden 'using namespace std' somewhere in the headers and it must be removed.
cc @eyonland

Comment (KalaivaniMCW) [reply ttnn/cpp/ttnn/operations/eltwise/binary/binary_composite.hpp]: will be addressed

**Review by eyonland — APPROVED** [review]

**Pull request merged**